### ISSUE TITLE
PET 2001-8N/2001 platform support possible stack pointer issue

### ISSUE THREAD

**Issue by voidstar78** [cc65/cc65#1986]
Observed that a PRG binary output that works for a 40XX PET often will not work for a 20XX-series PET.  This is as determined by dragging the generated binary (e.g. main.out) into VICE, with VICE configured with a Machine Type of either 4016/4016  vs  2001-8N.   An associated of mine does also have a physical working "wooden PET" (a 2001-series with the original wooden case).

The behavior was observed with the following simple program:

void main(void)
{	
  int x = 3;
	while (x > 0)   // works on            4016/4032 only
	{
		--x;
	}	
}

Even if the above happens to work for some user's configurations -- try adding additional declarations around the "int x", especially a mixture of types (like char and int) or try re-arranging the order of those declarations (e.g. "char a;  int x = 3;"), and try also turning off any optimization during the build.  This is to force a variety of alignment of the declarations.

The issue observed (on 20XX-series) is that this example does not "return to BASIC" and appears to go into an infinite loop.   In other cases, the result may be corrupting output (printed strings appear abnormal), or if it does return to BASIC some abnormal behavior is observed (like not being able to press ENTER) -- a sign of corrupted memory.     After some analysis, the following was noted:

- The pet.cfg defines ZP starting at $55 and going to $4A, with zeropage.s defining 26 bytes in that ZP region to store stackPointer and all the other registers
- After startup of the system and waiting to get to the initial BASIC prompt, using the VICE monitor it was observed that the state of this region of memory (from $55 to $55+$4A) are initialize/set to a slightly different state between the two system (20XX and 40XX). 
- It was noted that per documentation, the content of the (system) zeropage (i.e. $00 to $FF) does differ between the 20XX and 40XX series as notes by...
40XX:  www.zimmers.net/cbmpics/cbm/PETx/petmem.txt
20XX (1978 doc):  https://www.masswerk.at/pet/manuals/PET_User_Manual_%282001-8%29.pdf

After some discussion, then the following was speculated:  The interrupt vector on the 20XX may be behaving differently, or may be corrupting what cc65 is using as the zeropage.   Or more specifically, cc65 may initialize the StackPointer to $00FF  (may appear as $FF00 in the binary) and the 20XX ISR may also be conflicting with that address.  This is not confirmed -- but it was observed that on the 40XX series, the sample program above correctly and nicely returns the stackPointer to this value, whereas the 20XX-series "gets stuck" at around $A7FF as its value for the stackPointer.

NOTE: VICE shows the ROM files it is using when switching between 4032 and 2001-8N.  Our greatest desire would be for cc65 to make a single binary that can transparently work across both platforms (at least in terms of gracefully starting and exiting).  But we understand if there is a zeropage conflict between the two, then maybe separate PRG binaries may need to be maintained for these build targets.

It was also noted that the Studio PRG compiler may have a similar/same issue.  (PRG binaries produced there exhibit the same behavior of not being "backwards compatible" on 20XX systems)

Other notes:
- issue #1: " the IRQ interrupt vector is in different places on the earlier vs later PETs."... .'The contents of the interrupt vector in machines with the old ROMs is hexadecimal E66B, and in machines with the new ROMs hex E61B. ' The newer ROMs mess with $0090 (which is in the zeropage), but we're unsure of the old ROMs.
- Issue #2: the PIA interface to scan for keyboard codes is (possibly) different

"when the new ROMs came out, this is what people did to get old
games working on the new machines. eg The original Space Invaders game
that everyone played worked on the old machines with 8K of memory. As
new machines came out people wrote programs to look for the above
bytes and adjust the code." - the exact offset of patches are either lost to time (or were specific to that Space Invaders build)

**Comment by polluks** [comment]
You know "pet (all CBM PET systems except the 2001)", so support was always missing.
Good luck!

**Comment by mrdudz** [comment]
I cant imagine that it takes a lot of effort to make it work.... perhaps some tweaking of the crt0 is needed, and perhaps a couple library functions need to be fixed...

In any case, i think adding a separate target is the way to go, trying to produce binaries that work on both will require duplicating too much code in the end.

**Comment by voidstar78** [comment]
Well I would have just let it go, because I didn't think the 2001's could be very easily upgraded to 32K.  But turns out, they can.  And I didn't see a technical reason why the 20XX's couldn't support a C-compiler - so I was curious about it.  I'm not too fluent about cc65, but I just monitored the simplest C program I could get to replicate the issue for me (i.e. the resulting binary, and how they ran differently in VICE between its 20XX ROM's and 40XX ROM's).  I was pretty sure it wasn't a VICE issue.

If comes down to needing to relocate the 26-byte zeropage portion borrowed by cc65, or altering initial values of stack pointer or something -- then yea I think that forces it to being categorized as a new platform, in case existing software is using those assumptions.

All that said, as I recall the 20XX are inherently slower (not by CPU, but some issue in the CRT or extra interrupts or something?)  So folks may be disappointed in that result.

**Comment by mrdudz** [comment]
> If comes down to needing to relocate the 26-byte zeropage portion borrowed by cc65, or altering initial values of stack pointer or something -- then yea I think that forces it to being categorized as a new platform, in case existing software is using those assumptions.

I gave it a quick try last night - and it is more than that. A bunch of library functions (probably most things that call into the ROM) must be fixed, too.

**Comment by Rhialto** [comment]
Finding free space in the 2001's zero page should be easy: it has the 80 byte "BASIC input buffer" in it (at 10-90, according to the linked original manual) . It was very wasteful of Micro Soft to use up so many zero page locations, but it was very convenient for machine language programs. Later ROMs put that at $0200. I remember reading in magazines that people complained about losing their nice zero page scratch space :)

**Comment by mrdudz** [comment]
Yes, but that doesnt really fix anything. https://github.com/mrdudz/cc65/tree/addPet2001 contains a fixed linker config that does that - but the programs still dont work right :)

**Comment by Rhialto** [comment]
Indeed, unfortunately any use of low-memory or ROM addresses would need to be adjusted as well. Some of the jump table entries would be usable. I do remember that some of them aren't that useful, because they expect to be used from Basic and parse some Basic command text; I don't remember if that was already improved in Basic 2/4, or only later with the VIC-20.
I used to have a nice table for converting low-memory addresses, but that was probably a WordPro file or something like that and long since lost to history...
